src/main/java/com/spotify/dns/AbstractChangeNotifier.java

```diff
@@ -34,18 +36,25 @@
 
   private final AtomicReference<Listener<T>> listenerRef = new AtomicReference<Listener<T>>();
 
+  private final AtomicBoolean listenerNotified = new AtomicBoolean(false);
+
+  private ReentrantLock lock = new ReentrantLock();
```

Comment:
Why a ```ReentrantLock``` rather than simply using ```synchronized``` for ```setListener``` and ```fireRecordsUpdated``` ?

If we should use a ```ReentrantLock```, the member variable should be ```final```